README
fastCGI application for generate machine statistics.


debian/changelog
graph-generator (1.0.0-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- Cactus <medved_cool@mail.ru>  Sat, 28 Jan 2017 14:03:10 +0300


src/main.c
#include "chunks.h"

int main(int argc, char **argv)
{
	setbuf(stdout, NULL);
	setbuf(stderr, NULL);

	static struct argp_option options[] = {
	{"socket" ,				's',	"127.0.0.1:9000",	0,	"Socket addres with port",			0},
	{"threads" ,			't',	"4",				0,	"Count of using outputs threads",	0},
	{"file" ,				'f',	"/file/path",		0,	"File with graph paramethers",		0},
	{"offset" ,				'o',	"0",				0,	"Graph offset in pixels",			0},
	{"multiplier" ,			'm',	"1",				0,	"Multiplier for file paramether",	0},
	{"sleep" ,				'p',	"30",				0,	"Harvester thread sleep time",		0},
	{"fill" ,				'c',	"true",				0,	"Fill svg graph",					0},
	{0}};

	struct argp argp = {0};
	argp.options	= options;
	argp.parser		= parse_opt;
//	argp.doc		=	"Machine list format:"
//						"\n\nlogin:password@hostname\n"
//						"login2:password2@hostname2\n\n"
//						"Or login@hostname if you have public key authorisation. Each machine on a new line. Offline machines handled in main thread.";

	argp_parse (&argp, argc, argv, 0, 0, NULL);

	if(FILE_PATH == NULL)
	{
		fprintf(stderr, "ERROR: --file should have argument\n");
		exit(EXIT_FAILURE);
	}

	APP app;
	app.socket		= -1;
	app.count		= 60/6*24;
	app.offset		= OFFSET;
	app.multiplier	= MULTIPLIER;
	app.file_path	= FILE_PATH;
	app.sleep		= SLEEP;
	app.fill		= FILL;
	app.data		= calloc(app.count, sizeof(double));

	pthread_mutex_init(&(app.accept_mutex), NULL);

	FCGX_Init();

	app.socket = FCGX_OpenSocket(SOCKET, 20);
	if(app.socket < 0)
	{
		fputs("ERROR: Failed to open socket\n", stderr);
		return 1;
	}

	pthread_t* threads = calloc(THREAD_COUNT, sizeof(pthread_t));

	pthread_t harv;
	pthread_create(&harv, NULL, harv_thread, &app);

	for(size_t i = 0; i < THREAD_COUNT; i++)
		pthread_create(threads+i, NULL, out_thread, &app);

	for(size_t i = 0; i < THREAD_COUNT; i++)
		pthread_join(threads[i], NULL);

	free(app.data);
	free(threads);

	return 0;
}


src/chunks.h

#ifndef CHUNKS_H_
	#define CHUNKS_H_

	#include <stdlib.h>
	#include <stdio.h>
	#include <stdint.h>
	#include <string.h>
	#include <assert.h>
	#include <stddef.h>
	#include <time.h>
	#include <stdbool.h>

	#include <argp.h>
	#include <pthread.h>
	#include <fcgi_config.h>
	#include <fcgiapp.h>
	#include <unistd.h>
	#include <sys/stat.h>
	#include <fcntl.h>

	extern char* SOCKET;
	extern int THREAD_COUNT;
	extern char* FILE_PATH;
	extern double OFFSET;
	extern double MULTIPLIER;
	extern int SLEEP;
	extern bool FILL;

	#define GRAPH_PERIOD 6*60

	void* out_thread(void* ptr);
	void* harv_thread(void* ptr);
	void strcat_realloc(char** buff, char* str, intmax_t str_len, intmax_t* buff_size, intmax_t* buff_pos);
	void polyline_generate(char** buff, intmax_t* buff_size, double* graph, size_t count, double offset);
	size_t get_day_pos(size_t period_min);
	int parse_opt (int key, char* arg, struct argp_state* state);

	typedef struct APP
	{
		pthread_mutex_t accept_mutex;
		int socket;
		char* file_path;
		size_t sleep;
		double multiplier;
		double offset;
		bool fill;

		double* data;
		size_t count;
	} APP;

#endif /* CHUNKS_H_ */


src/polyline.c
#include "chunks.h"

#define SVG_HEADER "" \
				"<svg xmlns='http://www.w3.org/2000/svg' height='100%' width='100%' version='1.1' viewBox='0 0 270 193'>\r\n" \
				"	<g fill-opacity='0.8' transform='translate(20, 110) scale(1, -1)' stroke='#d07735' stroke-width='0.5' fill='#CCC'>\r\n"

#define SVG_POLYLINE_START ""\
				"		<polyline points='"
#define SVG_POLYLINE_END "'/>\r\n"
#define SVG_TIME_TEAMPLATE ""\
				"		<polyline stroke-opacity='0.3' fill='red' stroke='red' transform='translate(%d, 0)' points='0,0 1,-2 -1,-2 0,0 0,100'/>\r\n"
#define SVG_END ""\
				"	</g>\r\n"\
				"</svg>\r\n"


void polyline_point_add(char** buff, intmax_t* buff_size, intmax_t* buff_pos, size_t pos, double value)
{
	char point[16];
	snprintf(point, sizeof(point), "%d,%.3lf ", (int)pos, value);
	strcat_realloc(buff, point, strlen(point), buff_size, buff_pos);
}

void polyline_generate(char** buff, intmax_t* buff_size, double* graph, size_t count, double offset)
{
	assert(buff != NULL);
	assert(*buff != NULL);
	assert(buff_size != NULL);
	assert(graph != NULL);

	**buff				= '\0';
	intmax_t buff_pos	= 0;

	strcat_realloc(buff, SVG_HEADER, sizeof(SVG_HEADER)-1, buff_size, &buff_pos);	//todo fill
	strcat_realloc(buff, SVG_POLYLINE_START, sizeof(SVG_POLYLINE_START)-1, buff_size, &buff_pos);

	polyline_point_add(buff, buff_size, &buff_pos, 0, 0+offset);
	for(size_t i=0; i<count; i++)
	{
		polyline_point_add(buff, buff_size, &buff_pos, i , graph[i] + offset);
	}
	polyline_point_add(buff, buff_size, &buff_pos, count, 0+offset);

	strcat_realloc(buff, SVG_POLYLINE_END, sizeof(SVG_POLYLINE_END)-1, buff_size, &buff_pos);

	char time_buff[sizeof(SVG_TIME_TEAMPLATE) + 8];
	size_t day_pos = get_day_pos(GRAPH_PERIOD);
	snprintf(time_buff, sizeof(time_buff), SVG_TIME_TEAMPLATE, (int)day_pos);
	strcat_realloc(buff, time_buff, strlen(time_buff), buff_size, &buff_pos);

	strcat_realloc(buff, SVG_END, sizeof(SVG_END)-1, buff_size, &buff_pos);
}


src/out_thread.c
#include "chunks.h"

void* out_thread(void* ptr)
{
	APP* app						= (APP*) ptr;
	pthread_mutex_t* accept_mutex	= &(app->accept_mutex);

	FCGX_Request request;

	if(FCGX_InitRequest(&request, app->socket, 0) != 0)
	{
		fputs("ERROR: Can not init request\n", stderr);
		return NULL;
	}

	intmax_t buff_size	= 4096;
	char* buff			= malloc(buff_size);

	while(1)
	{
		pthread_mutex_lock(accept_mutex);
		int rc = FCGX_Accept_r(&request);
		pthread_mutex_unlock(accept_mutex);

		if(rc == 0)
		{
			polyline_generate(&buff, &buff_size, app->data, app->count, app->offset);
			FCGX_PutS("Content-type: image/svg+xml\r\n\r\n", request.out);
			FCGX_PutS(buff, request.out);
		}
		else
		{
			fputs("ERROR: Can not accept new request\n", stderr);
			break;
		}

		FCGX_Finish_r(&request);
	}

	return NULL;
}


src/argp.c
#include "chunks.h"

char* SOCKET			= "127.0.0.1:9000";
int THREAD_COUNT		= 2;
char* FILE_PATH			= NULL;
double OFFSET			= 0;
double MULTIPLIER		= 1.f;
int SLEEP				= 30;
bool FILL				= true;

int parse_opt(int key, char* arg, struct argp_state* state)
{
	switch (key)
	{
		case 's':
		{
			size_t arg_len = strlen(arg);
			if(arg_len != 0)
			{
				SOCKET = malloc(arg_len + 1);
				strncpy(SOCKET, arg, arg_len);
			}
			break;
		}
		case 't':
		{
			if(sscanf(arg, "%d", &THREAD_COUNT) != 1)
			{
				fprintf(stderr, "\nERROR: Failed to parse threads count %s\n", arg);
				exit(EXIT_FAILURE);
			}

			if(THREAD_COUNT > 64)
			{
				fprintf(stderr, "\nERROR: Too many threads. The maximum number of threads is 64\n");
				exit(EXIT_FAILURE);
			}
			break;
		}
		case 'f':
		{
			struct stat s;
			if(stat(arg, &s) == -1)
			{
				fprintf(stderr, "\nERROR: File %s is not exist\n", arg);
				exit(EXIT_FAILURE);
			}
			else
			{
				if(S_ISREG(s.st_mode) || S_ISCHR(s.st_mode))
				{
					size_t arg_len = strlen(arg);
					FILE_PATH = malloc(arg_len+1);
					strncpy(FILE_PATH, arg, arg_len);
				} else
				{
					fprintf(stderr, "\nERROR: %s is not a file\n", arg);
					exit(EXIT_FAILURE);
				}
			}
			break;
		}
		case 'o':
		{
			if(sscanf(arg, "%lf", &OFFSET) != 1)
			{
				fprintf(stderr, "\nERROR: Failed to parse offset %s\n", arg);
				exit(EXIT_FAILURE);
			}

			break;
		}
		case 'm':
		{
			if(sscanf(arg, "%lf", &MULTIPLIER) != 1)
			{
				fprintf(stderr, "\nERROR: Failed to parse multiplier %s\n", arg);
				exit(EXIT_FAILURE);
			}

			break;
		}
		case 'p':
		{
			if(sscanf(arg, "%d", &SLEEP) != 1)
			{
				fprintf(stderr, "\nERROR: Failed to parse sleep %s\n", arg);
				exit(EXIT_FAILURE);
			}

			break;
		}
		case 'c':
		{
			if(strcmp("true", arg) == 0)
				FILL = true;
			else if(strcmp("false", arg) == 0)
				FILL = false;
			else
			{
				fprintf(stderr, "\nERROR: Failed to parse fill %s\n", arg);
				exit(EXIT_FAILURE);
			}

			break;
		}
	}
	return 0;
}
